parachain/pallets/basic-channel/merkle-proof/src/lib.rs

```diff
@@ -0,0 +1,792 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
```

Comment:
Wondering if this should be here? Did we copy this file from Substrate?

Also asking because if it is, perhaps it's not necessary to review? 🤔

Reply:
We did 😬 It's from the `beefy-merkle-tree` crate in FRAME's beefy-mmr pallet, but with beefy-mmr's `Hash` replaced with the equivalent `sp_core::H256` (also a 32 byte array) and the `Hasher` trait replaced with `sp_runtime::traits::Hash` (a more general form of `Hasher`).

Come to think of it, the `mmr` pallet looks like it _mostly_ does what we need here. A big difference is that the `mmr` pallet runs on every block, while our channel only runs every few blocks, determined by the `interval`.

@vgeddes What do you think of trying to use the `mmr` pallet here? It looks to me like it doesn't quite fit, but it might be close enough for us to contribute a change to the pallet.

Reply:
Thanks for explaining! 👍